[postprocessor/embedthumbnail] Fix Youtube's WebP thumbnails by converting them to JPG

alexmerkel opened the pull request:
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

### Description of your *pull request* and other information

This change should fix #25687:
Youtube started using the [WebP](https://en.wikipedia.org/wiki/WebP) image format for thumbnail images. However, both `ffmpeg` and `AtomicParsley` only support the embedding of _JPEG_ and _PNG_ files. Therefore, as @noname8753 suggested, a conversion step was added during which the thumbnail is converted to _JPEG_ using `ffmpeg`.

empyrical commented:
This fix works for me! One thing that might need to be tweaked is to have it strip the original thumbnail extension when converting - right now it leaves a file `thumbnail.webp.jpg`, and it could instead name it `thumbnail.jpg`

alexmerkel commented:
Thanks @empyrical, I changed the file name.
Additionally I added a detection for WebP files that are mislabeled as JPG by Youtube (see @koloshor comment [here](https://github.com/ytdl-org/youtube-dl/issues/25687#issuecomment-647071373)).

BassThatHertz commented:
I added your "fix" and then ran youtube-dl with `python3 -m youtube-dl` but the same issue persists. Here's an example of the terminal output:

```[youtube] g15xwMv9A7I: Downloading webpage
[youtube] g15xwMv9A7I: Downloading thumbnail ...
[youtube] g15xwMv9A7I: Writing thumbnail to: NOIXES & Cemre Emin - Escape-g15xwMv9A7I.webp
[download] Destination: NOIXES & Cemre Emin - Escape-g15xwMv9A7I.webm
[download]   0.0% of 2.52MiB at 146.38KiB/s ETA 00:17
[download]   0.1% of 2.52MiB at 418.29KiB/s ETA 00:06
[download]   0.3% of 2.52MiB at 937.45KiB/s ETA 00:02
[download]   0.6% of 2.52MiB at  1.87MiB/s ETA 00:01
[download]   1.2% of 2.52MiB at 1013.49KiB/s ETA 00:02
[download]   2.4% of 2.52MiB at  1.30MiB/s ETA 00:01
[download]   4.9% of 2.52MiB at  1.68MiB/s ETA 00:01
[download]   9.9% of 2.52MiB at  2.45MiB/s ETA 00:00
[download]  19.8% of 2.52MiB at  3.31MiB/s ETA 00:00
[download]  39.6% of 2.52MiB at  4.36MiB/s ETA 00:00
[download]  79.3% of 2.52MiB at  5.23MiB/s ETA 00:00
[download] 100.0% of 2.52MiB at  5.39MiB/s ETA 00:00
[download] 100% of 2.52MiB in 00:00
[ffmpeg] Destination: NOIXES & Cemre Emin - Escape-g15xwMv9A7I.mp3
Deleting original file NOIXES & Cemre Emin - Escape-g15xwMv9A7I.webm (pass -k to keep)
[ffmpeg] Adding thumbnail to "NOIXES & Cemre Emin - Escape-g15xwMv9A7I.mp3"
ERROR: Conversion failed!```

alexmerkel commented:
> I cloned the repo, added your "fix" and then ran youtube-dl with `python3 -m youtube_dl` but the same issue persists.

Are you sure that you applied the patch correctly? I added a log message when a thumbnail conversion takes place and it is missing from your output.
This is the terminal output I get when running the same command as you with the patch applied:
```
$ python3 -m youtube_dl -x --embed-thumbnail --audio-format mp3 https://www.youtube.com/watch?v=gi9Bvj0JIrw
[youtube] gi9Bvj0JIrw: Downloading webpage
[youtube] gi9Bvj0JIrw: Downloading js player 1d33781a
[youtube] gi9Bvj0JIrw: Downloading js player 1d33781a
[youtube] gi9Bvj0JIrw: Downloading thumbnail ...
[youtube] gi9Bvj0JIrw: Writing thumbnail to: Mikebøi - Heavy Lungs (ft. SRU)-gi9Bvj0JIrw.webp
[download] Destination: Mikebøi - Heavy Lungs (ft. SRU)-gi9Bvj0JIrw.webm
[download] 100% of 3.71MiB in 00:01
[ffmpeg] Destination: Mikebøi - Heavy Lungs (ft. SRU)-gi9Bvj0JIrw.mp3
Deleting original file Mikebøi - Heavy Lungs (ft. SRU)-gi9Bvj0JIrw.webm (pass -k to keep)
[ffmpeg] Converting thumbnail "Mikebøi - Heavy Lungs (ft. SRU)-gi9Bvj0JIrw.webp" to JPEG
[ffmpeg] Adding thumbnail to "Mikebøi - Heavy Lungs (ft. SRU)-gi9Bvj0JIrw.mp3"
```

As you can see, there is an additional message `ffmpeg] Converting thumbnail "Mikebøi - Heavy Lungs (ft. SRU)-gi9Bvj0JIrw.webp" to JPEG` that is missing from your log.

BassThatHertz commented:
@alexmerkel I applied the fix correctly, the issue was that it wasn't running the version with the fix added, it was running the version I had previously installed using pip3. All is good now.

njoynj commented:
An error occurs when there is "%" in the file title

alexmerkel commented:
Oh, good catch @njoynj. `ffmpeg` interpreted `%` as an image sequence and got confused. I now replace `%` with `_` in the thumbnail file name, this fixes the issue.

po5 commented:
Please add a way to skip this, I do want webp thumbnails.

OzoneHelix commented:
I'd like to compile a binary of of Alex's patch but there is no instruction on how to compile it to a Windows exe binary it would be handy if someone wrote a wiki on how to do that but that is a side tangent

BassThatHertz commented:
For those who want to easily install youtube-dl with the fix implemented, you can do:
`pip install git+https://github.com/ytdl-org/youtube-dl.git@refs/pull/25717/head`

CoWinkKeyDinkInc commented:
I used the youtube-dl version that @BassThatHertz suggested and it works perfectly, after I cut out all the affected videos from my music playlist from `archive.txt`, it was able to overwrite all the old content and the thumbnails work! :+1:

Labud98 commented:
Hi, 
Here is my output with your patch.
Any ideas ? Thank you !

`StefSuse:/home/stefan # python3 -m youtube_dl --verbose --extract-audio --no-mtime --audio-quality 4 --embed-thumbnail --audio-format mp3 -o test.mp3 https://www.youtube.com/watch?v=Phsr8tC0R6w
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['--verbose', '--extract-audio', '--no-mtime', '--audio-quality', '4', '--embed-thumbnail', '--audio-format', 'mp3', '-o', 'test.mp3', 'https://www.youtube.com/watch?v=Phsr8tC0R6w']
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, pref UTF-8
[debug] youtube-dl version 2020.06.16.1
[debug] Python version 3.8.3 (CPython) - Linux-5.7.5-1-default-x86_64-with-glibc2.2.5
[debug] exe versions: ffmpeg 4.2.3, ffprobe 4.2.3
[debug] Proxy map: {}
[youtube] Phsr8tC0R6w: Downloading webpage
[youtube] {18} signature length 102, html5 player 3662280c
[youtube] {137} signature length 106, html5 player 3662280c
[youtube] {248} signature length 106, html5 player 3662280c
[youtube] {399} signature length 102, html5 player 3662280c
[youtube] {136} signature length 102, html5 player 3662280c
[youtube] {247} signature length 106, html5 player 3662280c
[youtube] {398} signature length 102, html5 player 3662280c
[youtube] {135} signature length 102, html5 player 3662280c
[youtube] {244} signature length 102, html5 player 3662280c
[youtube] {397} signature length 102, html5 player 3662280c
[youtube] {134} signature length 106, html5 player 3662280c
[youtube] {243} signature length 106, html5 player 3662280c
[youtube] {396} signature length 106, html5 player 3662280c
[youtube] {133} signature length 102, html5 player 3662280c
[youtube] {242} signature length 106, html5 player 3662280c
[youtube] {395} signature length 106, html5 player 3662280c
[youtube] {160} signature length 106, html5 player 3662280c
[youtube] {278} signature length 102, html5 player 3662280c
[youtube] {394} signature length 106, html5 player 3662280c
[youtube] {140} signature length 102, html5 player 3662280c
[youtube] {249} signature length 106, html5 player 3662280c
[youtube] {250} signature length 106, html5 player 3662280c
[youtube] {251} signature length 106, html5 player 3662280c
[youtube] Phsr8tC0R6w: Downloading thumbnail ...
[youtube] Phsr8tC0R6w: Writing thumbnail to: test.webp
[debug] Invoking downloader on 'https://r2---sn-1gi7znes.googlevideo.com/videoplayback?expire=1594708795&ei=2_4MX6-1D9XqgAeFqJTIAw&ip=84.74.236.91&id=o-AJekJH4GgIM7PnQSZ6hNb1XZj4YSe48shhQz_o81wfdH&itag=251&source=youtub                  e&requiressl=yes&mh=0N&mm=31%2C26&mn=sn-1gi7znes%2Csn-4g5ednld&ms=au%2Conr&mv=m&mvi=2&pl=19&initcwndbps=1981250&vprv=1&mime=audio%2Fwebm&gir=yes&clen=2888440&dur=159.181&lmt=1593246369017324&mt=1594687114&fvip=2&keepali                  ve=yes&c=WEB&txp=5531432&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cgir%2Cclen%2Cdur%2Clmt&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIgaQhX5jy3ADFJhGo3NRHcLm                  vQxMRCiEBIkAV0zI6OWR4CIQCpvlYBP0khzbQV3N-flbeg2DJOyTP_JWHUZvP6-ZBkfw%3D%3D&sig=AOq0QJ8wRQIgTPfKR0PlX_fsL4fF7lZvT7lGdVc3HKYozBCD2o9ieoUCIQDqyz-kaoP2miNF5IDW6uKLD09tJvROvTxTaoqDtVXVxw==&ratebypass=yes'
[download] test.mp3 has already been downloaded
[download] 100% of 2.93MiB
[debug] ffmpeg command line: ffprobe -show_streams file:test.mp3
[ffmpeg] Post-process file test.mp3 exists, skipping
[ffmpeg] Converting thumbnail "test.webp" to JPEG
[debug] ffmpeg command line: ffmpeg -y -loglevel repeat+info -i file:test.webp -bsf:v mjpeg2jpeg file:test.jpg
[ffmpeg] Adding thumbnail to "test.mp3"
[debug] ffmpeg command line: ffmpeg -y -loglevel repeat+info -i file:test.mp3 -i file:test.jpg -c copy -map 0 -map 1 -metadata:s:v 'title="Album cover"' -metadata:s:v 'comment="Cover (Front)"' file:test.temp.mp3
ERROR:   Stream #1:0 -> #0:1 (copy)
Traceback (most recent call last):
  File "/usr/local/lib/python3.8/site-packages/youtube_dl/YoutubeDL.py", line 2065, in post_process
    files_to_delete, info = pp.run(info)
  File "/usr/local/lib/python3.8/site-packages/youtube_dl/postprocessor/embedthumbnail.py", line 73, in run
    self.run_ffmpeg_multiple_files([filename, thumbnail_filename], temp_filename, options)
  File "/usr/local/lib/python3.8/site-packages/youtube_dl/postprocessor/ffmpeg.py", line 235, in run_ffmpeg_multiple_files
    raise FFmpegPostProcessorError(msg)
youtube_dl.postprocessor.ffmpeg.FFmpegPostProcessorError:   Stream #1:0 -> #0:1 (copy)`

adithya-s-sekhar commented:
Can confirm @Laburd98 's problem happens when you specify the output filename. 

Using output templates and not specifying a filename works fine.

adithya-s-sekhar commented:
More about the bug @Labud98 found

From original comment
> [download] 100% of 2.93MiB
> [debug] ffmpeg command line: ffprobe -show_streams file:test.mp3
> [ffmpeg] Post-process file test.mp3 exists, skipping

youtube-dl seems to save the downloaded opus file as test.mp3 and skips converting the audio to actual mp3 format. This breaks ffmpeg when it tries to look for mp3 streams in an opus file.

> Input #0, matroska,webm, from 'file:test.mp3':
>   Metadata:
>     encoder         : google/video-file
>   Duration: 00:02:39.18, start: -0.007000, bitrate: 145 kb/s
>     Stream #0:0(eng): Audio: opus, 48000 Hz, stereo, fltp (default)
> Input #1, image2, from 'file:test.jpg':
>   Duration: 00:00:00.04, start: 0.000000, bitrate: 18816 kb/s
>     Stream #1:0: Video: mjpeg (Baseline), yuvj420p(pc, bt470bg/unknown/unknown), 1280x720, 25 tbr, 25 tbn, 25 tbc
> [mp3 @ 0000024506284dc0] Invalid audio stream. Exactly one MP3 audio stream is required.
> Could not write header for output file #0 (incorrect codec parameters ?): Invalid argument

I'm not that good at python to fix this but hope this helps.

michaelb commented:
> StefSuse:/home/stefan # python3 -m youtube_dl --verbose --extract-audio --no-mtime --audio-quality 4 --embed-thumbnail --audio-format mp3 -o test.mp3 https://www.youtube.com/watch?v=Phsr8tC0R6w

Using the unpatched (unrelated to @alexmerkel PR) youtube-dl version, I can confirm **this  ^ is a different bug.**
Another issue should be opened.

(my output:`
youtube-dl --verbose --extract-audio --no-mtime --audio-quality 4 --embed-thumbnail --audio-format mp3 -o test.mp3 https://www.youtube.com/watch\?v\=Phsr8tC0R6w
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['--verbose', '--extract-audio', '--no-mtime', '--audio-quality', '4', '--embed-thumbnail', '--audio-format', 'mp3', '-o', 'test.mp3', 'https://www.youtube.com/watch?v=Phsr8tC0R6w']
[debug] Encodings: locale UTF-8, fs utf-8, out utf-8, pref UTF-8
[debug] youtube-dl version 2020.06.16.1
[debug] Python version 3.8.3 (CPython) - Linux-5.7.7-arch1-1-x86_64-with-glibc2.2.5
[debug] exe versions: ffmpeg 4.2.3, ffprobe 4.2.3, rtmpdump 2.4
[debug] Proxy map: {}
[youtube] Phsr8tC0R6w: Downloading webpage
[youtube] {18} signature length 107, html5 player 5253ac4d
[youtube] {137} signature length 107, html5 player 5253ac4d
[youtube] {248} signature length 107, html5 player 5253ac4d
[youtube] {399} signature length 103, html5 player 5253ac4d
[youtube] {136} signature length 107, html5 player 5253ac4d
[youtube] {247} signature length 107, html5 player 5253ac4d
[youtube] {398} signature length 107, html5 player 5253ac4d
[youtube] {135} signature length 107, html5 player 5253ac4d
[youtube] {244} signature length 107, html5 player 5253ac4d
[youtube] {397} signature length 107, html5 player 5253ac4d
[youtube] {134} signature length 107, html5 player 5253ac4d
[youtube] {243} signature length 107, html5 player 5253ac4d
[youtube] {396} signature length 107, html5 player 5253ac4d
[youtube] {133} signature length 107, html5 player 5253ac4d
[youtube] {242} signature length 107, html5 player 5253ac4d
[youtube] {395} signature length 103, html5 player 5253ac4d
[youtube] {160} signature length 107, html5 player 5253ac4d
[youtube] {278} signature length 107, html5 player 5253ac4d
[youtube] {394} signature length 103, html5 player 5253ac4d
[youtube] {140} signature length 107, html5 player 5253ac4d
[youtube] {249} signature length 107, html5 player 5253ac4d
[youtube] {250} signature length 107, html5 player 5253ac4d
[youtube] {251} signature length 107, html5 player 5253ac4d
[youtube] Phsr8tC0R6w: Downloading thumbnail ...
[youtube] Phsr8tC0R6w: Writing thumbnail to: test.webp
[debug] Invoking downloader on 'https://r2---sn-hgn7yn7e.googlevideo.com/videoplayback?expire=1594835573&ei=Fe4OX4i5C4mAxN8Pu9OM8A4&ip=2a01%3Acb1d%3A189%3A6600%3A9ead%3A97ff%3Afea3%3Ab285&id=o-ANEU-lylBuNe4Vf3lZ2agoxTtwatlCRmPiEOkzFdNRZy&itag=251&source=youtube&requiressl=yes&mh=0N&mm=31%2C26&mn=sn-hgn7yn7e%2Csn-4g5e6nze&ms=au%2Conr&mv=m&mvi=2&pl=30&initcwndbps=910000&vprv=1&mime=audio%2Fwebm&gir=yes&clen=2888440&dur=159.181&lmt=1593246369017324&mt=1594813833&fvip=2&keepalive=yes&beids=9466585&c=WEB&txp=5531432&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cgir%2Cclen%2Cdur%2Clmt&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRAIgSHVN0o5i4Gd267PtwR9JB43GuvytvK0txIgOqNoe-scCIBdtIfJT091FyozfDeUN2fjS89qEHi7-OI55KacG7OKT&sig=AOq0QJ8wRQIhAJnbED9VqaKixsvwjYVHRv2NgInaYiByjzOaFvqoBEUQAiAZqR_oy-I5S1c-8QZ7lx_1UfdHzCGD_ExnzdRVtgk1cQ==&ratebypass=yes'
[download] Destination: test.mp3
[download] 100% of 2.75MiB in 00:03
[debug] ffmpeg command line: ffprobe -show_streams file:test.mp3
[ffmpeg] Post-process file test.mp3 exists, skipping
[ffmpeg] Adding thumbnail to "test.mp3"
[debug] ffmpeg command line: ffmpeg -y -loglevel repeat+info -i file:test.mp3 -i file:test.webp -c copy -map 0 -map 1 -metadata:s:v 'title="Album cover"' -metadata:s:v 'comment="Cover (Front)"' file:test.temp.mp3
ERROR:   Stream #1:0 -> #0:1 (copy)
Traceback (most recent call last):
  File "/usr/lib/python3.8/site-packages/youtube_dl/YoutubeDL.py", line 2065, in post_process
    files_to_delete, info = pp.run(info)
  File "/usr/lib/python3.8/site-packages/youtube_dl/postprocessor/embedthumbnail.py", line 51, in run
    self.run_ffmpeg_multiple_files([filename, thumbnail_filename], temp_filename, options)
  File "/usr/lib/python3.8/site-packages/youtube_dl/postprocessor/ffmpeg.py", line 235, in run_ffmpeg_multiple_files
    raise FFmpegPostProcessorError(msg)
youtube_dl.postprocessor.ffmpeg.FFmpegPostProcessorError:   Stream #1:0 -> #0:1 (copy)
`
)



The PR looks clean enough to me.
Given how bad it breaks --embed-thumbnail for  a lot of youtube videos (every new one, if not all), can it be considered for merge?

Haafingar commented:
Youtube-dl is broken for a very large number of use cases because of this bug. No new issues with the PR have been found in 2 weeks, and it's pretty small and straightforward. I hate to be pushy, but why hasn't this been merged?

reallyuniquename commented:
Somewhat related: #24882

`ffmpeg` now supports thumbnail embedding into MP4 video files.

arc-arc commented:
batch downloaded ~80 YT videos as .MP3 files and used --embed-thumbnails . worked perfectly . videos were from various channels and upload dates. thanks for this fix!

yuceltoluyag commented:
ty save my life  @alexmerkel

Mhowser commented:
Could we get a merge please @dstftw 😀

silariel commented:
I just downloaded youtube-dl after youtube-dlg stopped working so I might be doing this wrong, but here's the output that I got with the same command:

```
youtube-dl -x --embed-thumbnail --audio-format mp3 https://www.youtube.com/watch?v=gi9Bvj0JIrw
[youtube] gi9Bvj0JIrw: Downloading webpage
[youtube] gi9Bvj0JIrw: Downloading thumbnail ...
[youtube] gi9Bvj0JIrw: Writing thumbnail to: Mikeb°i - Heavy Lungs (ft. SRU)-gi9Bvj0JIrw.webp
[download] Destination: Mikeb°i - Heavy Lungs (ft. SRU)-gi9Bvj0JIrw.webm
[download] 100% of 3.71MiB in 00:00
[ffmpeg] Destination: Mikeb°i - Heavy Lungs (ft. SRU)-gi9Bvj0JIrw.mp3
Deleting original file Mikeb°i - Heavy Lungs (ft. SRU)-gi9Bvj0JIrw.webm (pass -k to keep)
[ffmpeg] Converting thumbnail "Mikeb°i - Heavy Lungs (ft. SRU)-gi9Bvj0JIrw.webp" to JPEG
[ffmpeg] Adding thumbnail to "Mikeb°i - Heavy Lungs (ft. SRU)-gi9Bvj0JIrw.mp3"
```

It looks like it should work but the mp3 file still doesn't have the thumbnail embedded?

alexmerkel commented:
@silariel Unfortunately I can't reproduce your issue. When I run your command, the resulting file has the embedded thumbnail. Have you tried opening the MP3 with a different program to see if the thumbnail appears there?

BassThatHertz commented:
In what way are thumbnails added with youtube-dl added with youtube-dl different to, say, official cover of a song purchased from iTunes? Because I have noticed that the cover art added with youtube-dl does not show up in Groove Music and Windows Media Player but cover art of music that I have purchased does.

@silariel If you're on Windows, try foobar2000. It shows the cover art. I also like the fact that it shows the bitrate in realtime (at the bottom-right) for VBR files. See the image below. As you can see, it shows the cover art.

![image](https://user-images.githubusercontent.com/48166845/88407057-726ba800-cdc9-11ea-8bfc-d0d60c45a94a.png)

alexmerkel commented:
According to [this](https://stackoverflow.com/a/18718265) stackoverflow answer, it might be because the ffmpeg command uses `comment="Cover (Front)"` instead of `comment="Cover (front)"` and the `-id3v2_version 3` flag is missing. But I'm unable to test this as I don't have a Win machine handy atm.

BassThatHertz commented:
@alexmerkel I'm trying to test this myself by adding `-id3v2_version 3`. I'm guessing you simply need to change:
`self.run_ffmpeg(thumbnail_filename, jpg_thumbnail_filename, ['-bsf:v', 'mjpeg2jpeg'])`
to
`self.run_ffmpeg(thumbnail_filename, jpg_thumbnail_filename, ['-bsf:v', 'mjpeg2jpeg', 'id3v2_version', '3'])`?

I'm guessing the following error when running youtube-dl with `python -m` though: 

```
C:\Users\H\Desktop\youtube-dl>python -m youtube_dl -x --embed-thumbnail --audio-format mp3 --audio-quality 0 https://www.youtube.com/watch?v=Y25awNvNB4A -o C:/Users/H/Desktop/Test.mp3
[youtube] Y25awNvNB4A: Downloading webpage
[youtube] Y25awNvNB4A: Downloading MPD manifest
[youtube] Y25awNvNB4A: Downloading thumbnail ...
[youtube] Y25awNvNB4A: Writing thumbnail to: C:\Users\H\Desktop\Test.jpg
[dashsegments] Total fragments: 17
[download] Destination: C:\Users\H\Desktop\Test.mp3
[download] 100% of 2.56MiB in 00:02
[ffmpeg] Post-process file C:\Users\H\Desktop\Test.mp3 exists, skipping
[ffmpeg] Converting thumbnail "C:\Users\H\Desktop\Test.webp" to JPEG
[ffmpeg] Adding thumbnail to "C:\Users\H\Desktop\Test.mp3"
ERROR:   Stream #1:0 -> #0:1 (copy)
```

I get this error even without adding the `-id3v2_version 3`, the only change I made to embedthumbnail.py was changing it to your version of the file with the fix.

alexmerkel commented:
@BassThatHertz no, the flag has to be added to the ffmpeg command that adds the thumbnail to the MP3 file
```
options = [
    '-c', 'copy', '-map', '0', '-map', '1',
     '-metadata:s:v', 'title="Album cover"', '-metadata:s:v', 'comment="Cover (Front)"']
```
to
```
options = [
    '-c', 'copy', '-map', '0', '-map', '1', '-id3v2_version', '3',
     '-metadata:s:v', 'title="Album cover"', '-metadata:s:v', 'comment="Cover (front)"']
```

The other error should be the unrelated bug that occurs when specifying the output filename that was discussed [above](https://github.com/ytdl-org/youtube-dl/pull/25717#issuecomment-658729806).

BassThatHertz commented:
@alexmerkel Just did some testing. `Cover (Front)` does not need to be changed to a lowercase f, and adding `-id3v2_version 3` sorts the issue on Windows.

With regards to the output filename issue, I can see that someone else has brought up the same issue but nobody has explained what's causing that issue.

alexmerkel commented:
I'm not sure this should be added to this pull request.
While it certainly needs to be changed, it has nothing to do with the webp files Youtube introduced that this pull request addresses.
Therefore, I would propose creating a new pull request for the addition of the `-id3v2_version 3` flag.

BassThatHertz commented:
@alexmerkel Makes sense. I would issue a new pull request but the issue is that your fix hasn't been merged with the master branch yet, and adding a fix for cover art visibility on fix doesn't make sense because right now the thumbnails don't even get added to the file (due to being WebP).

alexmerkel commented:
@Labud98, @adithya-s-sekhar, @michaelb, @BassThatHertz I found the reason for the `ERROR:   Stream #1:0 -> #0:1 (copy)` bug, but it has nothing to do directly with the thumbnails.
When specifying an output file with a .mp3 extension, the conversion step from WEBM or M4A to MP3 is skipped because `ffmpeg` assumes the audio format is already correct.
In the logs this can be observed by the `[ffmpeg] Post-process file test.mp3 exists, skipping` message.
Additionally `ffprobe` shows the webm file format
```
Input #0, matroska,webm, from 'file:test.mp3':
  Metadata:
    encoder         : google/video-file
  Duration: 00:02:39.22, start: -0.007000, bitrate: 134 kb/s
    Stream #0:0(eng): Audio: opus, 48000 Hz, stereo, fltp (default)
```
So the file has the extension .mp3 but it isn't really an MP3 file.
The thumbnail-embedding process results in an exception because there `ffmpeg` is expecting an MP3 stream but is not receiving one and panics.

I think a new issue/pull request should be created to address this problem.

BassThatHertz commented:
@alexmerkel In which file/line is the extension from the filename specified with `-o` used by youtube-dl to set the name of the download? One solution is to change the extension from what the user entered to 'mka'. For example, if the user entered `file.mp3`, changing this to `file.mka` is one way of solving this issue. You can confirm that this will work by entering `-o test.mka` instead of `-o test.mp3` in the command line and there will be no error.

adithya-s-sekhar commented:
> When specifying an output file with a .mp3 extension, the conversion step from WEBM or M4A to MP3 is skipped because `ffmpeg` assumes the audio format is already correct.
> In the logs this can be observed by the `[ffmpeg] Post-process file test.mp3 exists, skipping` message.
> Additionally `ffprobe` shows the webm file format
> ...

I've noticed it [too](#issuecomment-658679812). This issue has nothing to do with the thumbnails.

BassThatHertz commented:
@Labud98, @adithya-s-sekhar, @michaelb, @alexmerkel I've submitted a fix for `ERROR: Stream #1:0 -> #0:1 (copy)` when specifying the mp3 extension for the output filename. [Here's ](https://github.com/ytdl-org/youtube-dl/pull/26110)the pull request.

BassThatHertz commented:
If possible, you should remove/decline my suggested changes. I don't want anything delaying this fix being merged to the master branch. If/when this fix gets merged, I may submit a pull request which passes `-id3v2_version 3` to FFmpeg so that the cover art can be seen when using Groove Music or Windows Media Player.

silariel commented:
> 
> 
> @BassThatHertz no, the flag has to be added to the ffmpeg command that adds the thumbnail to the MP3 file
> 
> ```
> options = [
>     '-c', 'copy', '-map', '0', '-map', '1',
>      '-metadata:s:v', 'title="Album cover"', '-metadata:s:v', 'comment="Cover (Front)"']
> ```
> 
> to
> 
> ```
> options = [
>     '-c', 'copy', '-map', '0', '-map', '1', '-id3v2_version', '3',
>      '-metadata:s:v', 'title="Album cover"', '-metadata:s:v', 'comment="Cover (front)"']
> ```
> 
> The other error should be the unrelated bug that occurs when specifying the output filename that was discussed [above](https://github.com/ytdl-org/youtube-dl/pull/25717#issuecomment-658729806).

Oops sorry, wasn't keeping track of this thread but this worked like a charm! :) ^ Thanks for the suggestions and help, y'all!

AvinashReddy3108 commented:
> Add `'-id3v2_version', '3'` to the `options` list to fix the issue of no cover art being seen when using Groove Music on Windows or Windows Media Player, and perhaps other applications.

Some older(?) Android devices/apps also needs the `'-write_id3v1', '1'` to be able to show the album art in their music players.

BassThatHertz commented:
@AvinashReddy3108 FFmpeg only supports ID3v2 versions 3 and 4.

TeguhTeknisi commented:
somehow, 'ffmpeg' in my Kubuntu 18.04 cannot convert webp.
but 'convert' from 'image magick' can convert it ( i need to 'apt install webp' first ), i hope you can make convert as option to convert the webp to jpg.

thanks.

AryanNateq commented:
I'm still having this issue as well when downloading mp4 videos on YouTube. Thumbnail downloads but doesn't embed.
`./youtube-dl https://youtu.be/WltJPKFo_J4 -f 137+140 -ciw --add-metadata --embed-subs --embed-thumbnail --write-sub --ffmpeg-location "C:\Program Files\ffmpeg\bin\ffmpeg.exe"`

dstftw commented on youtube_dl/postprocessor/embedthumbnail.py:
```diff
@@ -41,6 +41,28 @@ def run(self, info):
+        # If not a jpg or png thumbnail, convert it to jpg using ffmpeg
+        if not os.path.splitext(thumbnail_filename)[1].lower() in ['.jpg', '.png']:
+            jpg_thumbnail_filename = os.path.splitext(thumbnail_filename)[0] + ".jpg"
+            jpg_thumbnail_filename = os.path.join(os.path.dirname(jpg_thumbnail_filename), os.path.basename(jpg_thumbnail_filename).replace('%', '_'))  # ffmpeg interprets % as image sequence
```
`%d` in path still does not work:
```
[debug] System config: []
[debug] User config: []
[debug] Custom config: []
[debug] Command-line args: ['--embed-thu', 'https://www.youtube.com/watch?v=bSBTcQNPNqc', '-f', '18', '-v']
[debug] Encodings: locale cp1251, fs utf-8, out utf-8, pref cp1251
[debug] youtube-dl version 2020.09.06
[debug] Git HEAD: bf6c312
[debug] Python version 3.7.0 (CPython) - Windows-10-10.0.10240-SP0
[debug] exe versions: ffmpeg N-85653-gb4330a0, ffprobe N-85653-gb4330a0, phantomjs 2.1.1, rtmpdump 2.4
[debug] Proxy map: {}
[youtube] bSBTcQNPNqc: Downloading webpage
[youtube] bSBTcQNPNqc: Downloading thumbnail ...
[youtube] bSBTcQNPNqc: Writing thumbnail to: How to print %d using printf in C language-bSBTcQNPNqc.jpg
[debug] Invoking downloader on 'https://r3---sn-ug5onuxaxjvh-v8cl.googlevideo.com/videoplayback?expire=1600046769&ei=UXJeX7bGEY-Hv_IP_o6dqAc&ip=109.171.17.81&id=o-ANtCWU23BMMI6vhhPbpRo2IK5dtAC86InaQPv-3vD5-I&itag=18&source=youtube&requiressl=yes&mh=qB&mm=31%2C29&mn=sn-ug5onuxaxjvh-v8cl%2Csn-ug5onuxaxjvh-n8v6&ms=au%2Crdu&mv=m&mvi=3&pl=17&initcwndbps=1216250&vprv=1&mime=video%2Fmp4&gir=yes&clen=8799654&ratebypass=yes&dur=512.185&lmt=1486031782975981&mt=1600025074&fvip=4&c=WEB&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cvprv%2Cmime%2Cgir%2Cclen%2Cratebypass%2Cdur%2Clmt&sig=AOq0QJ8wRQIgDa1K2jtiyU_gu1OFQJt90jbHXlUVOlfrsa5M4Lvg8DgCIQDAvZFUNWRzXJj3MVs85ma_uiLTcYa6v0T6sU7D-n94vg%3D%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRAIgK7n5iyTpA0A3QixL9srxxZeDGR29UvLruC6e_S8xPbYCIEz-R679FFbiG1Fm2v7cb0E148lrISI_BCP_EE611v7K'
[download] How to print %d using printf in C language-bSBTcQNPNqc.mp4 has already been downloaded
[download] 100% of 8.39MiB
[ffmpeg] Converting thumbnail "How to print %d using printf in C language-bSBTcQNPNqc.webp" to JPEG
[debug] ffmpeg command line: ffmpeg -y -loglevel "repeat+info" -i "file:How to print %d using printf in C language-bSBTcQNPNqc.webp" "-bsf:v" mjpeg2jpeg "file:How to print _d using printf in C language-bSBTcQNPNqc.jpg"
ERROR: file:How to print %d using printf in C language-bSBTcQNPNqc.webp: No such file or directory
Traceback (most recent call last):
  File "C:\Dev\youtube-dl\master\youtube_dl\YoutubeDL.py", line 2065, in post_process
    files_to_delete, info = pp.run(info)
  File "C:\Dev\youtube-dl\master\youtube_dl\postprocessor\embedthumbnail.py", line 61, in run
    self.run_ffmpeg(thumbnail_filename, jpg_thumbnail_filename, ['-bsf:v', 'mjpeg2jpeg'])
  File "C:\Dev\youtube-dl\master\youtube_dl\postprocessor\ffmpeg.py", line 239, in run_ffmpeg
    self.run_ffmpeg_multiple_files([path], out_path, opts)
  File "C:\Dev\youtube-dl\master\youtube_dl\postprocessor\ffmpeg.py", line 235, in run_ffmpeg_multiple_files
    raise FFmpegPostProcessorError(msg)
youtube_dl.postprocessor.ffmpeg.FFmpegPostProcessorError: file:How to print %d using printf in C language-bSBTcQNPNqc.webp: No such file or directory
```

dstftw commented on youtube_dl/postprocessor/embedthumbnail.py:
```diff
@@ -41,6 +41,28 @@ def run(self, info):
+
+        # If not a jpg or png thumbnail, convert it to jpg using ffmpeg
+        if not os.path.splitext(thumbnail_filename)[1].lower() in ['.jpg', '.png']:
+            jpg_thumbnail_filename = os.path.splitext(thumbnail_filename)[0] + ".jpg"
```
`utils.replace_extension`.
